resources/matrix_irc.rb

```diff
@@ -0,0 +1,121 @@
+unified_mode true
+
+default_action :create
+
+property :container_name, String, name_property: true
+property :config, Hash, default: {}
+property :port, Integer, default: 8090
+property :host_domain, String, required: true
+property :host_name, String, default: lazy { "synapse-#{host_domain}" }
+property :host_path, String, default: lazy { "/opt/#{host_name}" }
+property :key_appservice, String, default: lazy { osl_matrix_genkey(host_name + container_name) }
+property :key_homeserver, String, default: lazy { osl_matrix_genkey(container_name + host_name) }
+property :tag, String, default: 'latest'
+property :sensitive, [true, false], default: true
+
+action :create do
+  # Pull down the latest version
+  docker_image 'matrixdotorg/matrix-appservice-irc' do
+    tag new_resource.tag
+  end
+
+  # Generate the app service file
+  osl_synapse_appservice(
+    'matrix-appservice-irc',
+    "http://#{new_resource.container_name}:#{new_resource.port}",
+    new_resource.key_appservice,
+    new_resource.key_homeserver,
+    'appservice-irc',
+    {
+      'users' => [
+        {
+          'exclusive' => true,
+          'regex' => '@as-irc_.*',
+        },
+      ],
+      'rate_limited' => false,
+      'protocols' => [
+        'irc',
+      ],
+    }
+  )
+
+  # Generate the config file
+  file "#{new_resource.host_path}/#{new_resource.container_name}-config.yaml" do
+    content osl_matrix_irc_defaults(new_resource.config)
+    owner 'synapse'
+    group 'synapse'
+    mode '400'
+    sensitive true
+  end
+
+  # Generate the passkey, one time
+  execute 'Generating Matrix-Appservice-IRC Passkey' do
+    command "openssl genpkey -out \"#{new_resource.host_path}/keys/irc-passkey.pem\" -outform PEM -algorithm RSA -pkeyopt rsa_keygen_bits:2048; chmod 400 '#{new_resource.host_path}/keys/irc-passkey.pem'"
+    user 'synapse'
+    group 'synapse'
+    creates "#{new_resource.host_path}/keys/irc-passkey.pem"
+  end
+
+  # Generate the signingkey, one time
+  # While we will NOT be using the media proxy, the bridge still requires this file.
+  # Using docker_container always throws a 400 HTTP error, executing docker works fine. Review requested.
+  # ---
+  #  docker_container 'Generate signingkey' do
+  #    repo 'matrixdotorg/matrix-appservice-irc'
+  #    autoremove true
+  #    volume "#{new_resource.host_path}/keys:/data"
+  #    entrypoint 'sh'
+  #    command ['-c', 'node lib/generate-signing-key.js > /data/signingkey.jwk && chmod 400 /data/signingkey.jwk']
+  #    user osl_synapse_user
+  #    not_if { ::File.exist?("#{new_resource.host_path}/keys/irc-signingkey.jwk") }
+  #  end
+  execute 'Generate signingkey' do
+    command "docker run --rm --entrypoint \"sh\" --volume #{new_resource.host_path}/keys:/data --user 994:989 matrixdotorg/matrix-appservice-irc \"-c\" \"node lib/generate-signing-key.js > /data/signingkey.jwk && chmod 400 /data/signingkey.jwk\""
+    creates "#{new_resource.host_path}/keys/signingkey.jwk"
+  end
```

Comment:
Is this something we could generate manually and just have it in a databag item instead? Same for the passkey above?

Reply:
I mean, we could. I was just trying to minimize the amount of static files/databags we needed to add in order to make these resources self-contained. These two resources are ran only one time, so it shouldn't be a worry for slowing down the scheduled converges.

As for the passkey, `osl_hookshot` also runs the same command, so there is already precedent for that.

As for the signing key here, we do not utilize it at all, so it's also possible to hardcode a file into the program, if speed is an issue.

Reply:
Okay, you've documented this well enough that I think we can keep it in. If we run into issues moving forward, we can rethink that.